[CSS] New property page for animation-composition (#19848)
mdn/content: files/en-us/web/css/animation-composition/index.md

---
title: animation-composition
slug: Web/CSS/animation-composition
tags:
  - CSS
  - CSS Animations
  - CSS Property

>   - Reference

Author: Yeah, I would leave this because we'll be getting rid of tags

  - recipe:css-property
browser-compat: css.properties.animation-composition
---
{{CSSRef}}

> The **`animation-composition`** [CSS](/en-US/docs/Web/CSS) property specifies the composite operation to use when multiple animations affect the same property simultaneously.

Author: Thanks for pointing this - I'm adding a glossary page for 'composite operation'.

The spec itself redirects to https://drafts.csswg.org/web-animations-1/#composite-operation for the term. This term in glossary will also be useful for [KeyframeEffect.composite](https://developer.mozilla.org/en-US/docs/Web/API/KeyframeEffect/composite).

> The **`animation-composition`** [CSS](/en-US/docs/Web/CSS) property specifies the composite operation to use when multiple animations affect the same property simultaneously.

Reviewer: Excellent. thanks.

<!-- It is often convenient to use the shorthand property {{cssxref("animation")}} to set all animation properties at once.-->

## Syntax

```css
/* Single animation */
animation-composition: replace;
animation-composition: add;
animation-composition: accumulate;

/* Multiple animations */
animation-composition: replace, add;
animation-composition: add, accumulate;
animation-composition: replace, add, accumulate;

/* Global values */
animation-composition: inherit;
animation-composition: initial;
animation-composition: revert;
animation-composition: revert-layer;
animation-composition: unset;
```

> **Note:** When you specify multiple comma-separated values on an `animation-*` property, they will be assigned to the animations specified in the {{cssxref("animation-name")}} property in different ways depending on how many there are. For more information, see [Setting multiple animation property values](/en-US/docs/Web/CSS/CSS_Animations/Using_CSS_animations#setting_multiple_animation_property_values).

### Values

- `replace`

>   - : The effect value overrides the underlying value of the property. This is the default composite operation.

Reviewer: ```suggestion
  - : The animated value overrides any non-`!important` underlying values of the property. This is the default value.
```

Author: We do need to use the terms "effect value" and "underlying value"...

- `add`

>   - : The effect value builds on the underlying value of the property. This values produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.

Reviewer: ```suggestion
  - : The animated value builds on the underlying value of the property. This value produces an additive effect. For property types where it is not possible to add multiple values, the animated value overrides the non-!important non-animated value. 
```

Author: I am going to retain the terms "effect value" and "underlying value". I have tried to explain them in the "Description" section.
I'll need to cross-check if the edits to the sentence "For property types..." convey the intended meaning here.

>   - : The effect value builds on the underlying value of the property. This values produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.

Author: I do not know the answer to this.

Not sure if this will help or is related, but I had found this earlier:

> "The most important thing to be aware of is that properties that can be typically accelerated, such as "opacity", "filter", "transform", "translate", "scale" and "rotate" will not be fully accelerated if composite operations are used. Instead, the animation resolution will operate on the main thread, while the application of the property will operate on a GraphicsLayer. We are continuing to look into how to address this in the future.

(source: https://bugs.webkit.org/show_bug.cgi?id=189299)

- `accumulate`

>   - : The effect and underlying values are combined. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.

Author: - I used the word 'combined' here and not 'added' to distinguish from the other operation named `add`. So to describe `add`, I used 'builds on' and for `accumulate`, I used 'combined', just a synonym that is not the word add.
- The word 'commutative' has not been defined or linked in the spec and I am not sure if they really mean a simple add, so I retained commutative.
- From what I understand, the order between the `underlying value` and `effect value` is important in a case like this (was shared by the engineer):

```css
@keyframes anim {
  from { transform: translateX(50px); }
  to { ... }
}
#target {
  transform: translateX(100px) rotate(90deg);
}
```

Assume that the `animation-composition` specified in `#target` is `add`.
The underlying value for `transform` here is `translateX(100px) rotate(90deg)` and the effect value is `translateX(50px)`

Since `add` combines underlying value and then effect value, so the effective "from" value will be `transform: translateX(100px) rotate(90deg) translateX(50px)` and not `transform: translateX(50px) translateX(100px) rotate(90deg)`.
It is not really the order of animations but how the final effect value is calculated.

Let me know if you were trying to make a different point.

>   - : The effect and underlying values are combined. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.

Reviewer: your original text is fine. I am being too detailed. We can add all the detail when it's actually supported and we can see what is going on in the debugger (right now, dev tools aren't showing animation tools when composition is set)

## Description

Each property that is targeted by the [@keyframes](/en-US/docs/Web/CSS/@keyframes) at-rule is associated with an effect stack. The value of the effect stack is calculated by combining the _underlying value_ of a property in a CSS style rule with the _effect value_ of that property in the keyframe. The `animation-composition` property helps to specify how to combine the underlying value with the effect value.

For example, in the CSS below, `blur(5px)` is the underlying value, and `blur(10px)` is the effect value. The `filter` property is affected by both animations, `pulse` and `brightness-pulse`. The `animation-composition` property specifies the operation to perform to produce the final effect value after compositing the effect of the underlying value and the effect value.

```css
.icon:hover {
  filter: blur(5px);

>   animation: brightness-pulse 3s infinite;

Reviewer: i found this error in the specification, so PR'ed a fix. That said, I think our example should be different from the spec so users can see a few valid use cases rather than one (currently invalid) repeated

>   animation: brightness-pulse 3s infinite;

Author: If we do specify the two animations like so:

>   animation: brightness 3s infinite alternate, pulse 2s infinite alternate;

Then I'd also have to add another `keyframes` at-rule for brightness, correct? There's only one right now for pulse. 

The other alternative is to update the 'animation' property to match `@keyframes pulse`:

>   animation: pulse 3s infinite alternate;

>   animation: brightness-pulse 3s infinite;

Reviewer: The `animation-name` did not match the `@keyframes <name>`. The merged the PR. I thought you were doing two animations, 1: brightness, and 2: pulse. But you and the spec authors were just doing one.

>   animation: brightness-pulse 3s infinite;

Author: I just updated the animation shorthand to place the keyframe name at the end: `animation: 3s infinite alternate pulse;`

  animation-composition: add;
}

@keyframes pulse {
  0% {
    filter: blur(10px);
  }
  10% {
    filter: blur(20px);
  }
}
```

Consider different values for the `animation-composition` property in the above example. The final effect value in each of those cases will be calculated as explained below:

- With `replace`, `blur(10px)` will replace `blur(5px)` in the `0%` keyframe. This is the default behavior of the property.

> - With `add`, the composite effect value in `0%` keyframe will be `blur(5px) blur(10px)`.

Reviewer: ```suggestion
- With `add`, the composite effect value in the `0%` keyframe will be `blur(5px) blur(10px)`.
```

> - With `add`, the composite effect value in `0%` keyframe will be `blur(5px) blur(10px)`.

Reviewer: ```suggestion
- With `add`, the composite effect value in the `0%` keyframe will be `blur(5px) blur(10px)`.
```

- With `accumulate`, the composite effect value in `0%` keyframe will be `blur(15px)`.

> **Note:** A composite operation can also be specified in a keyframe. In that case, the specified composite operation is used for each property first within that keyframe and then on each property in the next keyframe.

## Formal definition

{{cssinfo}}

## Formal syntax

{{csssyntax}}

## Examples

### Understanding the animation-composition values

The example below shows the effect of different `animation-composition` values side-by-side.

#### HTML

```html
<div class="container">replace
  <div id="replace" class="target"></div>
</div>
<div class="container">add
  <div id="add" class="target"></div>
</div>
<div class="container">accumulate
  <div id="accumulate" class="target"></div>
</div>
```

#### CSS

Here the underlying value is `translateX(50px) rotate(90deg)`.

```css
@keyframes slide {

>   from {transform: translateX(100px);}

Reviewer: ```suggestion
  20%, 40% {transform: translateX(100px);}
```

Author: I've made the suggested change but I feel it is adding more complexity to the use case. There will be too many things to unpack.

>   to {transform: translateX(150px);}

Reviewer: ```suggestion
  80%, 100% {transform: translateX(150px);}
```

}
.container {
  width: 240px;
  height: 220px;
  background: cyan;
  display: inline-block;
}
.target {
  width: 20px;
  height: 50px;
  background: green;
  border-radius: 10px;

>   transform: translateX(50px) rotate(90deg);

Reviewer: ```suggestion
  transform:  translateX(50px) rotate(45deg) ;
```

It is a little more obvious what is happening to translateX at 45deg than 90deg, because 90 is vertical.

>   animation: slide 2s linear infinite;

Reviewer: ```suggestion
  animation: slide 5s linear infinite;
```

Provide a little more time so people can grok the effect. Also, I used 20/40/80/100% so stuff lasts 1s

> }

Reviewer: ```suggestion
}
.target:hover {
   animation-play-state: paused;
 }
```

Reviewer: Enable pausing on hover so users can see the effect, especially of to color.

Reviewer: before adding the color explanation, let's see if the other edits make sense. The color additives are cool. we should likely explain how the r, g, and b values are added or replaced if that is in fact what is happening. i have gone down this rabbit hole too long today...

#replace {
  animation-composition: replace;
}
#add {
  animation-composition: add;
}
#accumulate {
  animation-composition: accumulate;
}
```

#### Result

{{EmbedLiveSample("Reversing the animation direction","100%","250")}}

> - With `replace`, the composite effect value in the `from` keyframe is `transform: translateX(100px)` (replacing `translateX(50px) rotate(90deg)`).

Reviewer: ```suggestion
- With `replace`, the composite effect value in the `20%` keyframe is `transform: translateX(100px)` (completely replacing `translateX(50px) rotate(45deg)`).
```

Author: This is great, thanks!

> - With `add`, the composite effect value in the `from` keyframe is `translateX(50px) rotate(90deg)` followed by `transform: translateX(100px)`.

Reviewer: ```suggestion
- With `add`, the composite effect value in the `20%` keyframe is `translateX(50px) rotate(45deg)` followed by `transform: translateX(100px)`.
```

this still doesn't make sense to me. You can't have two translate values at a time. That doesn't mean it's not possible, I am just not grokking it.

> - With `accumulate`, the composite effect value in the `from` keyframe is `translateX(150px) rotate(90deg)`.

Reviewer: ```suggestion
- With `accumulate`, the composite effect value in the `20%` keyframe is `translateX(150px) rotate(45deg)`.
```

---
title: animation-composition
slug: Web/CSS/animation-composition
tags:
  - CSS
  - CSS Animations
  - CSS Property
  - Reference
  - recipe:css-property
browser-compat: css.properties.animation-composition
---
{{CSSRef}}

> The **`animation-composition`** [CSS](/en-US/docs/Web/CSS) property specifies the composite operation to use when multiple animations affect the same property simultaneously.

Reviewer: ```suggestion
The **`animation-composition`** CSS property specifies the [composite operation](/en-US/docs/glossary/composite_operation) to use when multiple animations affect the same property simultaneously.
```

<!--the interactive example will be created in a separate PR. Will uncomment the macro below when I have an interactive example in place. Note that this property is still behind a flag in FF and not yet implemented in Safari and Chrome.-->
<!-- {{EmbedInteractiveExample("pages/css/animation-composition.html")}} -->

<!--only longhand animation-composition is enabled in FF104, so commenting out the sentence below for now. Not removing the sentence because it can be uncommented when support in shorthand animation property is added.-->

> <!-- It is often convenient to use the shorthand property {{cssxref("animation")}} to set all animation properties at once.-->

Reviewer 2: I think these comments can be removed and we can create an issue or a TODO somewhere to track this

Author: Yes! Thanks for pointing these out. I've removed the comments. For future tracking, I've added it to our 'to triage' list.

animation-composition: replace;
animation-composition: add;
animation-composition: accumulate;

/* Multiple animations */
animation-composition: replace, add;
animation-composition: add, accumulate;
animation-composition: replace, add, accumulate;

/* Global values */
animation-composition: inherit;
animation-composition: initial;
animation-composition: revert;
animation-composition: revert-layer;
animation-composition: unset;
```

> **Note:** When you specify multiple comma-separated values on an `animation-*` property, they will be applied to the animations in the order in which the {{cssxref("animation-name")}}s appear. If the number of animations and compositions differ, the values listed in the `animation-composition` property will cycle from the first to the last `animation-name`, looping until all the animations have an assigned `animation-composition` value. For more information, see [Setting multiple animation property values](/en-US/docs/Web/CSS/CSS_Animations/Using_CSS_animations#setting_multiple_animation_property_values).

### Values

- `replace`
  - : The effect value overrides the underlying value of the property. This is the default value.
- `add`

>   - : The effect value builds on the underlying value of the property. This operation produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.

Reviewer: ```suggestion
  - : The effect value builds on the underlying value of the property. This operation produces an additive effect. For animation types where the addition operation is not commutative, the order of the operands is the underlying value followed by the effect value.
```

- `accumulate`
  - : The effect and underlying values are combined. For animation types where the addition operation is not commutative, the order of the operands is underlying value followed by the effect value.

## Description

Each property that is targeted by the [@keyframes](/en-US/docs/Web/CSS/@keyframes) at-rule is associated with an effect stack. The value of the effect stack is calculated by combining the _underlying value_ of a property in a CSS style rule with the _effect value_ of that property in the keyframe. The `animation-composition` property helps to specify how to combine the underlying value with the effect value.

For example, in the CSS below, `blur(5px)` is the underlying value, and `blur(10px)` is the effect value. The `filter` property is affected by both animations, `pulse` and `brightness-pulse`. The `animation-composition` property specifies the operation to perform to produce the final effect value after compositing the effect of the underlying value and the effect value.

```css
.icon:hover {
  filter: blur(5px);

>   animation: brightness-pulse 3s infinite;

Reviewer: ```suggestion
  animation: pulse 3s infinite;
```

- With `replace`, `blur(10px)` will replace `blur(5px)` in the `0%` keyframe. This is the default behavior of the property.
- With `add`, the composite effect value in the `0%` keyframe will be `blur(5px) blur(10px)`.
- With `accumulate`, the composite effect value in `0%` keyframe will be `blur(15px)`.

> **Note:** A composite operation can also be specified in a keyframe. In that case, the specified composite operation is used for each property first within that keyframe and then on each property in the next keyframe.

## Formal definition

{{cssinfo}}

## Formal syntax

{{csssyntax}}

## Examples

### Understanding the animation-composition values

The example below shows the effect of different `animation-composition` values side-by-side.

#### HTML

```html
<div class="container">replace
  <div id="replace" class="target"></div>
</div>
<div class="container">add
  <div id="add" class="target"></div>
</div>
<div class="container">accumulate
  <div id="accumulate" class="target"></div>
</div>
```

#### CSS

Here the underlying value is `translateX(50px) rotate(45deg)`.

```css
@keyframes slide {

>   20%, 40% {

Reviewer: ```suggestion
  0%, 20% {
```

Reviewer: i am reverting the change i made a few days ago. We need to have it at 0% because the content is talking about what happens at the 0% mark. we should still hold it there of the 20% so people can see the difference between the 3.

Author: The explanation text for this example does talk about the '20%, 40%' keyframe. This is in the 'Examples' section.
The one with the '0%' keyframe is a different example in the 'Description' section. 
So I'm keeping this as is `20%, 40%`.

Reviewer: the reason to change it back to 0% is so it is additive / cumulative from the start. If we start at 20%, then the 0% is the original value, and the effect is less visible. I made a mistake when I suggested changing it to 20%

Author: fixing this here: https://github.com/mdn/content/pull/20250
